[WIP] RHDEVDOCS-3917 - documents new authentication methods for remote_write in monitoring (#44713)
openshift/openshift-docs: modules/monitoring-supported-remote-write-authentication-settings.adoc

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: REFERENCE
[id="supported_remote_write_authentication_settings_{context}"]
= Supported remote write authentication settings

You can use different methods to authenticate with a remote write endpoint.

> Currently supported authentication methods are AWS Signature Version 4, basic authentication, bearer token authorization, OAuth 2.0, and TLS client.

Reviewer: ```suggestion
Currently supported authentication methods are AWS Signature Version 4, basic authentication, authorization, OAuth 2.0, and TLS client.
```

Author: deleted

The following table provides details about supported authentication methods for use with remote write.

[options="header"]
|===

|Authentication method|Config map field|Description

|AWS Signature Version 4|`sigv4`|This method uses AWS Signature Version 4 authentication to sign requests.

> You cannot use this method simultaneously with bearer token, OAuth 2.0, or basic authentication.

Reviewer 2: ```suggestion
You cannot use this method simultaneously with a bearer token, OAuth 2.0, or basic authentication.
```

Reviewer: ```suggestion
You cannot use this method simultaneously with authorization, OAuth 2.0, or basic authentication.
```

Author: fixed

|basic authentication|`basicAuth`|Basic authentication sets the authorization header on every remote write request with the configured username and password.

> |bearer token authorization|`Bearer`|Bearer token authentication sets the authorization header on every remote write request using the configured token.

Reviewer: IIUC the correct entry in the configmap is `authorization` (see below in the examples).

Reviewer: ```suggestion
|authorization|`authorization`|Authorization sets the Authorization header on every remote write request using the configured token.
```

Author: Edited per your suggestion

> You can configure bearer token authentication by specifying a secret that contains the bearer token information. 
> When you use this method, the `remoteWrite` settings in the config map must include an `authorization` `type` of `Bearer`.

Reviewer: ```suggestion
```

Author: Deleted

|OAuth 2.0|`oauth2`|An OAuth 2.0 configuration uses the client credentials grant type. 

> Prometheus fetches an access token from `tokenUrl` with the specified client ID and client secret to access the remote write endpoint using this token.

Reviewer 2: ```suggestion
Prometheus fetches an access token from `tokenUrl` with the specified client ID and client secret to access the remote write endpoint.
```

Author: deleted the final phrase per your suggestion

> You cannot use this method simultaneously with bearer token, AWS Signature Version 4, or basic authentication.

Reviewer: ```suggestion
You cannot use this method simultaneously with authorization, AWS Signature Version 4, or basic authentication.
```

Author: replaced per your suggestion

|TLS client|`tlsConfig`|A TLS client configuration specifies the CA certificate, the client certificate, and the client key file information used to authenticate with the remote write endpoint server using TLS.
The sample configuration assumes that you have already created a CA certificate file, a client certificate file, and a client key file.

|===

== Config map location for authentication settings
The following shows the location of the authentication configuration in the `ConfigMap` object for default platform monitoring.

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write-endpoint.example.com" <1>
        <endpoint_authentication_details> <2>
----
<1> The URL of the remote write endpoint.
<2> The required configuration details for the authentication method for the endpoint.

> Currently supported authentication methods are Amazon Web Services (AWS) Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 

Reviewer: ```suggestion
Currently supported authentication methods are Amazon Web Services (AWS) Signature Version 4, authorization, basic authentication, OAuth 2.0, and TLS client. 
```

Author: replaced

[NOTE]
====

> If you want to configure remote write for the Prometheus instance that monitors user-defined projects, edit the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace.

Reviewer 2: ```suggestion
If you configure remote write for the Prometheus instance that monitors user-defined projects, edit the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace.
```

Author: Fixed and added a sentence: `Note that the Prometheus config map component is called `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object and not `prometheusK8s`, as it is in the `cluster-monitoring-config` `ConfigMap` object. `

====

== Example remote write authentication settings

The following samples show different authentication settings you can use to connect to a remote write endpoint.
Each sample also shows how to configure a corresponding `Secret` object that contains authentication credentials and other relevant settings.
Each sample configures authentication for use with default platform monitoring in the `openshift-monitoring` namespace.

.Sample YAML for AWS Signature Version 4 authentication

The following shows the settings for a `sigv4` secret named `sigv4-credentials` in the `openshift-monitoring` namespace.

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: sigv4-credentials
  namespace: openshift-monitoring
stringData:

>   access: <AWS_access_key> <1>
>   secret: <AWS_secret_key> <2>

Reviewer: `secret` might be confused with the Kubernetes Secret resource. Maybe this?
```suggestion
  accessKey: <AWS_access_key> <1>
  secretKey: <AWS_secret_key> <2>
```

type: Opaque
----
<1> The AWS API access key.
<2> The AWS API secret key.

The following shows sample AWS Signature Version 4 remote write authentication settings that use a `Secret` object named `sigv4-credentials` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://authorization.example.com/api/write"
        sigv4:
          region: <AWS_region> <1>
          accessKey:
            name: sigv4-credentials <2>
            key: access <3>
          secretKey:
            name: sigv4-credentials <2>

>             key: secret <3>

Reviewer 2: Should we represent any of the values in the preceding five lines be between angle brackets?

Author: No. These are the actual values. They correspond to the fields in the `Secret` object that contain the credentials.

          profile: <AWS_profile_name> <4>
          roleArn: <AWS_role_arn> <5>
----
<1> The AWS region.
<2> The name of the `Secret` object containing the AWS API access credentials.

> <3> The AWS API access credentials contained in the specified `Secret` object.

Reviewer: To be consistent with the rest of the examples I would expect each field to have its own footnote.
```suggestion
<3> The key in the specified `Secret` object containing the AWS access key.
<4> The key in the specified `Secret` object containing the AWS secret key.
```

Author: Added the extra calloout/footnote and edited per your suggestion as follows:

```
<3> The key that contains the AWS API access key in the specified `Secret` object.
<4> The key that contains teh AWS API secret key in the specified `Secret` object.
<5> The name of the AWS profile that is being used to authenticate.
<6> The unique identifier for the Amazon Resource Name (ARN) assigned to your role.
```

.Sample YAML for basic authentication

The following shows sample basic authentication settings for a `Secret` object named `rw-basic-auth` in the `openshift-monitoring` namespace: 

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: rw-basic-auth
  namespace: openshift-monitoring
stringData:
  user: <basic_username> <1>
  password: <basic_password> <2>
type: Opaque
----
<1> The username.
<2> The password.

The following sample shows a `basicAuth` remote write configuration that uses a `Secret` object named `rw-basic-auth` in the `openshift-monitoring` namespace.
It assumes that you have already set up authentication credentials for the endpoint. 

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://basicauth.example.com/api/write"
        basicAuth:
          username:
            name: rw-basic-auth <1>
            key: user <2>
          password:
            name: rw-basic-auth <1>
            key: password <3>
----
<1> The name of the `Secret` object that contains the authentication credentials.

> <2> The username contained in the specified `Secret` object.
> <3> The password contained in the specified `Secret` object.

Reviewer: ```suggestion
<2> The key in the specified `Secret` object containing the username.
<3> The key in the specified `Secret` object containing the password.
```

Author: Edited per your suggestion as follows:

```
<2> The key that contains the username  in the specified `Secret` object.
<3> The key that contains the password in the specified `Secret` object.
```

> .Sample YAML for bearer token authentication using a `Secret` Object

Reviewer: ```suggestion
.Sample YAML for authentication with `Authorization` header using a `Secret` Object
```

Author: changed to your later suggestion

> The following shows bearer token settings for a `Secret` object named `rw-bearer-auth` in the `openshift-monitoring` namespace:

Reviewer: I think that it's fine to mention bearer token from this point (to keep the example concrete).

Author: @simonpasquier Would it be more accurate to say `bearer token authorization` rather than `bearer token authentication`?

Reviewer: The configuration entry is called `authorization` because this is the name of the HTTP header but this is misleading since the mechanism is really meant to authenticate the HTTP client (see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization). So while it's `authorization` in the configuration file, I believe that the documentation should use authentication/authenticate.

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: rw-bearer-auth
  namespace: openshift-monitoring
stringData:
  token: <authentication_token> <1>
type: Opaque
----
<1> The authentication token.

The following shows sample bearer token config map settings that use a `Secret` object named `rw-bearer-auth` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
    prometheusK8s:
      remoteWrite:
      - url: "https://authorization.example.com/api/write"
        authorization:

>           type: Bearer

Reviewer: ```suggestion
          type: Bearer <x>      
 ...
 <x> Authentication type of the request (default: `Bearer`).
```

Author: Added the following callout per your suggestion:

`<1> The authentication type of the request. The default value is `Bearer`.
`

          credentials:
            name: rw-bearer-auth <1>
            key: token <2>
----
<1> The name of the `Secret` object that contains the authentication credentials.

> <2> The authentication token contained in the specified `Secret` object.

Reviewer: ```suggestion
<2> The key in the specified `Secret` object containing the authentication token.
```

Author: Edited per your suggestion as follows:

`<2> The key that contains the authentication token in the specified `Secret` object.
`

  token: <oauth2_authentication_token> <3>
type: Opaque
----
<1> The Oauth 2.0 ID.
<2> The OAuth 2.0 secret.
<3> The OAuth 2.0 token.

The following shows an `oauth2` remote write authentication sample configuration that uses a `Secret` object named `oauth2-credentials` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://test.example.com/api/write"
        oauth2:
          clientId:
            secret:
              name: oauth2-credentials <1>
              key: id <2>
          clientSecret:
            name: oauth2-credentials <1>
            key: secret <2>
          tokenUrl: https://example.com/oauth2/token <3>
          scopes: <4>
          - <scope_1>
          - <scope_2>
          endpointParams: <5>
            param1: <parameter_1>
            param2: <parameter_2>
----
<1> The name of the corresponding `Secret` object. Note that `ClientId` can alternatively refer to a `ConfigMap` object, although `clientSecret` must refer to a `Secret` object.

> <2> The OAuth 2.0 credentials contained in the specified `Secret` object.

Reviewer: ```suggestion
<2> The key in the specified `Secret` object containing the OAuth 2.0 credentials.
```

Author: Edited per your suggestion as follows:

`<2> The key that contains the OAuth 2.0 credentials in the specified `Secret` object.
`

> <3> URL used to fetch a token with the specified `clientId` and `clientSecret`.

Reviewer 2: ```suggestion
<3> The URL used to fetch a token with the specified `clientId` and `clientSecret`.
```

Author: Fixed

  client.crt: <client_cert> <2>
  client.key: <client_key> <3>
type: tls
----
<1> The CA certificate in the Prometheus container with which to validate the server certificate. 
<2> The client certificate for authentication with the server.
<3> The client key.

The following sample shows a `tlsConfig` remote write authentication configuration that uses a TLS `Secret` object named `mtls-bundle`.

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write-endpoint.example.com"
        tlsConfig:
          ca:
            secret:
              name: mtls-bundle <1>
              key: ca.crt <2>
          cert:
            secret:
              name: mtls-bundle <1>
              key: client.crt <3>
          keySecret:
            name: mtls-bundle <1>
            key: client.key <4>
----
<1> The name of the corresponding `Secret` object that contains the TLS authentication credentials. Note that `ca` and `cert` can alternatively refer to a `ConfigMap` object, though `keySecret` must refer to a `Secret` object.

> <2> The CA certificate for the endpoint as configured in the specified `Secret` object.

Reviewer 2: ```suggestion
<2> The CA certificate for the endpoint, as configured in the specified `Secret` object.
```

> <3> The client certificate for the endpoint as configured in the specified `Secret` object.

Reviewer 2: ```suggestion
<3> The client certificate for the endpoint, as configured in the specified `Secret` object.
```

Author: fixed

Each sample configures authentication for use with default platform monitoring in the `openshift-monitoring` namespace.

.Sample YAML for AWS Signature Version 4 authentication

The following shows the settings for a `sigv4` secret named `sigv4-credentials` in the `openshift-monitoring` namespace.

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: sigv4-credentials
  namespace: openshift-monitoring
stringData:
  access: <AWS_access_key> <1>
  secret: <AWS_secret_key> <2>
type: Opaque
----
<1> The AWS API access key.
<2> The AWS API secret key.

The following shows sample AWS Signature Version 4 remote write authentication settings that use a `Secret` object named `sigv4-credentials` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://authorization.example.com/api/write"
        sigv4:
          region: <AWS_region> <1>
          accessKey:
            name: sigv4-credentials <2>

>             key: access

Author: Good idea. I added in callout footnotes for all of the key fields in all of the code samples

            key: access
          secretKey:
            name: sigv4-credentials <2>
            key: secret
          profile: <AWS_profile_name> <3>
          roleArn: <AWS_role_arn> <4>
----
<1> The AWS region.
<2> The name of the `Secret` object containing the AWS API access credentials.
<3> The name of the AWS profile that is being used to authenticate.
<4> The unique identifier for the Amazon Resource Name (ARN) assigned to your role.

> .Sample for basic authentication

Author: fixed

The following shows sample basic authentication settings for a `Secret` object named `rw-basic-auth` in the `openshift-monitoring` namespace: 

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: rw-basic-auth
  namespace: openshift-monitoring
stringData:
  user: <basic_username> <1>
  password: <basic_password> <2>
type: Opaque
----
<1> The username.
<2> The password.

The following sample shows a `basicAuth` remote write configuration that uses a `Secret` object named `rw-basic-auth` in the `openshift-monitoring` namespace.
It assumes that you have already set up authentication credentials for the endpoint. 

[source,yaml]

> ----

Author: good point. Fixed

kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://basicauth.example.com/api/write"
        basicAuth:
          username:
            name: rw-basic-auth <1>
            key: user
          password:
            name: rw-basic-auth <1>
            key: password
----
<1> The name of the `Secret` object that contains the authentication credentials.

> .Sample YAML for bearer token authentication using a token file

Reviewer: as I commented above, I believe that we only need an example for `bearer`.

Author: deleted

The following shows a sample bearer token configuration that points to a token file. 
It assumes that a bearer token file exists in the location specified for `bearerTokenFile`.

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://authorization.example.com/api/write"
        bearerTokenFile: "<path_to_token_file>" <1>
----
<1> Path to the bearer token file.

> .Sample for bearer token authentication using a `Secret` Object

Author: fixed

The following shows bearer token settings for a `Secret` object named `rw-bearer-auth` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: rw-bearer-auth
  namespace: openshift-monitoring
stringData:
  token: <authentication_token> <1>
type: Opaque
----
<1> The authentication token.

> The following shows sample bearer token config map settings that use a `Secret` object named `rw-bearer-auth`in the `openshift-monitoring` namespace:

Author: fixed

  secret: <oauth2_secret> <2>
  token: <oauth2_authentication_token> <3>
type: Opaque
----
<1> The Oauth 2.0 ID.
<2> The OAuth 2.0 secret.
<3> The OAuth 2.0 token.

The following shows an `oauth2` remote write authentication sample configuration that uses a `Secret` object named `oauth2-credentials` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://test.example.com/api/write"
        oauth2:
          clientId:
            secret:
              name: oauth2-credentials <1>
              key: id
          clientSecret:
            name: oauth2-credentials <1>
            key: secret
          tokenUrl: https://example.com/oauth2/token <2>
          scopes: <3>
          - <scope_1>
          - <scope_2>
          endpointParams: <4>
            param1: <value_1>
            param2: <value_2>
----
<1> The name of the corresponding `Secret` object. Note that `ClientId` can alternatively refer to a `ConfigMap` object, though `clientSecret` must refer to a `Secret` object.

> <2> URL used to access a token with the specified `clientId` and `clientSecret`.

Author: changed `access` to `fetch`

<3> The OAuth 2.0 scopes for the authorization request. These scopes limit what data the tokens can access.
<4> The OAuth 2.0 authorization request parameters required for the authorization server.

.Sample YAML for TLS client authentication

The following shows sample TLS client settings for a `tls` `Secret` object named `selfsigned-mtls-bundle` in the `openshift-monitoring` namespace.

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:

>   name: selfsigned-mtls-bundle

Author: deleted `selfsigned-` throughout

Prometheus fetches an access token from `tokenUrl` with the specified client ID and client secret to access the remote write endpoint using this token.
You cannot use this method simultaneously with bearer token, AWS Signature Version 4, or basic authentication.

|TLS client|`tlsConfig`|A TLS client configuration specifies the CA certificate, the client certificate, and the client key file information used to authenticate with the remote write endpoint server using TLS.
The sample configuration assumes that you have already created a CA certificate file, a client certificate file, and a client key file.

|===

== Config map location for authentication settings
The following shows the location of the authentication configuration in the `ConfigMap` object.

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write-endpoint.example.com" <1>
        <endpoint_authentication_details> <2>
----
<1> The URL of the remote write endpoint.
<2> The required configuration details for the authentication method for the endpoint.

> Currently supported authentication methods are Amazon Web Services (AWS) Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 

Reviewer 3: Maybe here it also makes sense to add the note about the UWM ConfigMap just for good measure, WDYT?

Author: Thanks @JoaoBraveCoding. That's a good idea. I've added the following admonition here:

```
[NOTE]
====
If you want to configure remote write for the Prometheus instance that monitors user-defined projects, edit the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. Note that TLS client configuration for remote write is not supported for use with user-defined projects.
====
```

        authorization:
          type: Bearer
          credentials:
            name: rw-bearer-auth <1>
            key: token
----
<1> The name of the `Secret` object that contains the authentication credentials.

.Sample YAML for OAuth 2.0 authentication

The following shows sample OAuth 2.0 settings for a `Secret` object named `oauth2-credentials` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: oauth2-credentials
  namespace: openshift-monitoring
stringData:
  id: <oauth2_id> <1>
  secret: <oauth2_secret> <2>
  token: <oauth2_authentication_token> <3>
type: Opaque
----
<1> The Oauth 2.0 ID.
<2> The OAuth 2.0 secret.
<3> The OAuth 2.0 token.

The following shows an `oauth2` remote write authentication sample configuration that uses a `Secret` object named `oauth2-credentials` in the `openshift-monitoring` namespace:

[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://test.example.com/api/write"
        oauth2:
          clientId:

>             secret:

Reviewer 4: shall we mention the fields such as `clientId` can either be a secret or configmap?
same for the fields in tls config: `ca`, `cert`. 
`clientSecret` and `keySecret` must be secret though.
